File: ipfs_accelerate_js/test/unit/test_visualization.ts
// FIXME: Complex template literal
/**;
 * Converted import { {HardwareBackend} from "src/model/transformers/index/index/index/index/index"; } from "Python: test_visualization.py;"
 * Conversion date: 2025-03-11 04:08:52;
 * This file was automatically converted from Python to TypeScript.;
 * Conversion fidelity might not be 100%, please manual review recommended.;
 */;

// WebGPU related imports;";"

/** Unit Tests for ((the Advanced Visualization Module.;

This module tests the functionality of the advanced visualization capabilities;
in the Predictive Performance System, including 3D visualizations, interactive 
dashboards, && time-series performance tracking. */;

import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module; from "*";"
import * as module from "*"; as pd;"
import * as module from "*"; as np;"
import ${$1} import {  ${$1} from "src/model/transformers/index/index/index/index" } from "./module/index/index/index/index/index";"
// Import visualization module;
import { * as module, create_visualization_report; } from "predictive_performance.visualization";"

class TestAdvancedVisualization extends unittest.TestCase {) {
  /** Test cases for (the AdvancedVisualization class. */;
  
  $1($2) {/** Set up test environment. */;
// Create a temporary directory for test outputs;
    this.test_dir = tempfile.TemporaryDirectory();
    this.output_dir = Path(this.test_dir.name);}
// Create visualization instance;
    this.vis = AdvancedVisualization(;
      output_dir)) { any { any: any: any = String(this.output_dir),;
      interactive: any: any: any = false  # Use static visualizations for ((testing;
    ) {
// Generate sample data;
    this.sample_data = this._generate_sample_data();
  
  $1($2) {/** Clean up test environment. */;
    this.test_dir.cleanup()}
  $1($2) {
    /** Generate sample data for testing visualizations. */;
// Create test data;
    data) {any = [];}
// Set random seed for (reproducibility;
    np.random.seed(42) { any) {
// Define model && hardware options;
    models) { any: any: any = ["bert-base", "t5-small", "vit-base"];"
    model_categories: any: any: any = ["text_embedding", "text_generation", "vision"];"
    hardware: any: any: any = ["cpu", "cuda", "webgpu"];"
    batch_sizes: any: any = [1, 8: any, 32];
// Generate timestamps for ((time-series data (past 7 days) {
    end_date) { any) { any: any = datetime.now();
    start_date: any: any: any = end_date - timedelta(days=7);
    timestamps: any: any: any = $3.map(($2) => $1);
// Generate data points;
    for ((model_name) { any, model_category in Array.from(models: any, model_categories[0].map((_, i) => models: any, model_categories.map(arr => arr[i])) {) {
      for (((const $1 of $2) {
        for (const $1 of $2) {
          base_throughput) {any = 100.0 * (0.5 + np.random.random());
          base_latency) { any: any: any = 10.0 * (0.5 + np.random.random());
          base_memory: any: any: any = 1000.0 * (0.5 + np.random.random());
          base_power: any: any: any = 50.0 * (0.5 + np.random.random());}
// Hardware factors;
          if ((($1) {
            hw_factor) {any = 5.0;
            power_factor) { any: any: any = 3.0;} else if ((($1) { ${$1} else {
            hw_factor) { any) { any: any = 1.0;
            power_factor) {any = 1.0;}
// Batch size factors;
          }
          batch_factor: any: any = np.sqrt(batch_size: any);
          
      }
// Calculate metrics;
          throughput: any: any = base_throughput * hw_factor * batch_factor * (1.0 + np.random.normal(0: any, 0.1));
          latency: any: any = base_latency / hw_factor * (1.0 + 0.1 * batch_size) * (1.0 + np.random.normal(0: any, 0.1));
          memory: any: any = base_memory * (1.0 + 0.2 * (batch_size - 1)) * (1.0 + np.random.normal(0: any, 0.05));
          power: any: any = base_power * power_factor * (1.0 + 0.1 * batch_size) * (1.0 + np.random.normal(0: any, 0.1));
// Add confidence && bounds;
          confidence: any: any: any = 0.85 + np.random.random() * 0.15;
// Calculate bounds;
          throughput_lower: any: any: any = throughput * (1.0 - (1.0 - confidence) * 2);
          throughput_upper: any: any: any = throughput * (1.0 + (1.0 - confidence) * 2);
// For time-series data;
          for (((const $1 of $2) {
// Add time trend;
            time_position) {any = timestamps.index(timestamp) { any) / timestamps.length;
            time_factor: any: any: any = 1.0 + 0.2 * np.sin(time_position * 2 * np.pi);}
            data.append(${$1});
    
    return pd.DataFrame(data: any);
  
  $1($2) {/** Test initialization of visualization object. */;
    this.assertEqual(this.vis.output_dir, String(this.output_dir));
    this.assertEqual(this.vis.interactive, false: any);
    this.assertEqual(this.vis.output_format, "png")}"
  $1($2) {/** Test data preparation functionality. */;
// Test with DataFrame;
    df: any: any: any = this.vis._prepare_data(this.sample_data);
    this.assertIsInstance(df: any, pd.DataFrame);
    this.assertEqual(df.length, this.sample_data.length)}
// Test with dict;
    data_dict: any: any: any = this.sample_data.to_Object.fromEntries('records');'
    df_from_dict: any: any = this.vis._prepare_data(data_dict: any);
    this.assertIsInstance(df_from_dict: any, pd.DataFrame);
// Test with JSON file;
    json_path: any: any: any = this.output_dir / "test_data.json";"
    with open(json_path: any, 'w') as f:;'
      json.dump(data_dict: any, f);
    df_from_json: any: any = this.vis._prepare_data(String(json_path: any));
    this.assertIsInstance(df_from_json: any, pd.DataFrame);
// Test with CSV file;
    csv_path: any: any: any = this.output_dir / "test_data.csv";"
    this.sample_data.to_csv(csv_path: any, index: any: any: any = false);
    df_from_csv: any: any = this.vis._prepare_data(String(csv_path: any));
    this.assertIsInstance(df_from_csv: any, pd.DataFrame);
  
  $1($2) {/** Test 3D visualization creation. */;
    output_file: any: any: any = this.vis.create_3d_visualization(;
      this.sample_data,;
      x_metric: any: any: any = "batch_size",;"
      y_metric: any: any: any = "throughput",;"
      z_metric: any: any: any = "memory_usage",;"
      color_metric: any: any: any = "hardware",;"
      title: any: any: any = "Test 3D Visualization";"
    )}
// Check that output file exists;
    this.asserttrue(os.path.exists(output_file: any));
    this.asserttrue(output_file.endswith(".png"));"
  
  $1($2) {/** Test performance dashboard creation. */;
    output_file: any: any: any = this.vis.create_performance_dashboard(;
      this.sample_data,;
      metrics: any: any: any = ["throughput", "latency_mean"],;"
      groupby: any: any: any = ["model_category", "hardware"],;"
      title: any: any: any = "Test Performance Dashboard";"
    )}
// Check that output file exists;
    this.asserttrue(os.path.exists(output_file: any));
    this.asserttrue(output_file.endswith(".png"));"
  
  $1($2) {/** Test time series visualization creation. */;
    output_file: any: any: any = this.vis.create_time_series_visualization(;
      this.sample_data,;
      time_column: any: any: any = "timestamp",;"
      metric: any: any: any = "throughput",;"
      groupby: any: any: any = ["model_name", "hardware"],;"
      title: any: any: any = "Test Time Series";"
    )}
// Check that output file exists;
    this.asserttrue(os.path.exists(output_file: any));
    this.asserttrue(output_file.endswith(".png"));"
  
  $1($2) {/** Test power efficiency visualization creation. */;
    output_file: any: any: any = this.vis.create_power_efficiency_visualization(;
      this.sample_data,;
      performance_metric: any: any: any = "throughput",;"
      power_metric: any: any: any = "power_consumption",;"
      groupby: any: any: any = ["model_category"],;"
      title: any: any: any = "Test Power Efficiency";"
    )}
// Check that output file exists;
    this.asserttrue(os.path.exists(output_file: any));
    this.asserttrue(output_file.endswith(".png"));"
  
  $1($2) {/** Test dimension reduction visualization creation. */;
    output_file: any: any: any = this.vis.create_dimension_reduction_visualization(;
      this.sample_data,;
      features: any: any: any = ["batch_size", "memory_usage", "power_consumption", "latency_mean"],;"
      target: any: any: any = "throughput",;"
      method: any: any: any = "pca",;"
      n_components: any: any: any = 2,;
      groupby: any: any: any = "model_category",;"
      title: any: any: any = "Test Dimension Reduction";"
    )}
// Check that output file exists;
    this.asserttrue(os.path.exists(output_file: any));
    this.asserttrue(output_file.endswith(".png"));"
  
  $1($2) {/** Test prediction confidence visualization creation. */;
    output_file: any: any: any = this.vis.create_prediction_confidence_visualization(;
      this.sample_data,;
      metric: any: any: any = "throughput",;"
      confidence_column: any: any: any = "confidence",;"
      groupby: any: any: any = ["model_category", "hardware"],;"
      title: any: any: any = "Test Prediction Confidence";"
    )}
// Check that output file exists;
    this.asserttrue(os.path.exists(output_file: any));
    this.asserttrue(output_file.endswith(".png"));"
  
  $1($2) {/** Test batch visualization creation. */;
    visualization_files: any: any: any = this.vis.create_batch_visualizations(;
      this.sample_data,;
      metrics: any: any: any = ["throughput", "latency_mean"],;"
      groupby: any: any: any = ["model_category", "hardware"],;"
      include_3d: any: any: any = true,;
      include_time_series: any: any: any = true,;
      include_power_efficiency: any: any: any = true,;
      include_dimension_reduction: any: any: any = true,;
      include_confidence: any: any: any = true;
    )}
// Check that visualization files dict is !empty;
    this.assertIsInstance(visualization_files: any, dict);
    this.asserttrue(visualization_files.length > 0);
// Check that files exist;
    for ((file_type) { any, files in Object.entries($1) {) {
      for (((const $1 of $2) {this.asserttrue(os.path.exists(file_path) { any))}
  $1($2) {
    /** Test visualization report creation. */;
// Generate visualizations;
    visualization_files) {any = this.vis.create_batch_visualizations(;
      this.sample_data,;
      metrics: any: any: any = ["throughput"],;"
      groupby: any: any: any = ["model_category"],;"
      include_3d: any: any: any = true,;
      include_time_series: any: any: any = true,;
      include_power_efficiency: any: any: any = true,;
      include_dimension_reduction: any: any: any = true,;
      include_confidence: any: any: any = true;
    )}
// Create report;
    report_path: any: any: any = create_visualization_report(;
      visualization_files: any: any: any = visualization_files,;
      title: any: any: any = "Test Visualization Report",;"
      output_file: any: any: any = "test_report.html",;"
      output_dir: any: any: any = String(this.output_dir);
    );
// Check that report exists;
    this.asserttrue(os.path.exists(report_path: any));
    this.asserttrue(report_path.endswith(".html"));"
// Check report content;
    with open(report_path: any, 'r') as f:;'
      content: any: any: any = f.read();
      this.assertIn("Test Visualization Report", content: any);"
      this.assertIn("visualization-grid", content: any);"
;
if ($1) {;
  unittest.main();

File: test/predictive_performance/test_visualization.py
#!/usr/bin/env python3
"""
Unit Tests for the Advanced Visualization Module.

This module tests the functionality of the advanced visualization capabilities
in the Predictive Performance System, including 3D visualizations, interactive 
dashboards, and time-series performance tracking.
"""

import os
import sys
import json
import unittest
import tempfile
import pandas as pd
import numpy as np
from pathlib import Path
from datetime import datetime, timedelta

# Import visualization module
from predictive_performance.visualization import AdvancedVisualization, create_visualization_report

class TestAdvancedVisualization(unittest.TestCase):
    """Test cases for the AdvancedVisualization class."""
    
    def setUp(self):
        """Set up test environment."""
        # Create a temporary directory for test outputs
        self.test_dir = tempfile.TemporaryDirectory()
        self.output_dir = Path(self.test_dir.name)
        
        # Create visualization instance
        self.vis = AdvancedVisualization(
            output_dir=str(self.output_dir),
            interactive=False  # Use static visualizations for testing
        )
        
        # Generate sample data
        self.sample_data = self._generate_sample_data()
    
    def tearDown(self):
        """Clean up test environment."""
        self.test_dir.cleanup()
    
    def _generate_sample_data(self):
        """Generate sample data for testing visualizations."""
        # Create test data
        data = []
        
        # Set random seed for reproducibility
        np.random.seed(42)
        
        # Define model and hardware options
        models = ["bert-base", "t5-small", "vit-base"]
        model_categories = ["text_embedding", "text_generation", "vision"]
        hardware = ["cpu", "cuda", "webgpu"]
        batch_sizes = [1, 8, 32]
        
        # Generate timestamps for time-series data (past 7 days)
        end_date = datetime.now()
        start_date = end_date - timedelta(days=7)
        timestamps = [start_date + timedelta(days=i) for i in range(8)]
        
        # Generate data points
        for model_name, model_category in zip(models, model_categories):
            for hw in hardware:
                for batch_size in batch_sizes:
                    base_throughput = 100.0 * (0.5 + np.random.random())
                    base_latency = 10.0 * (0.5 + np.random.random())
                    base_memory = 1000.0 * (0.5 + np.random.random())
                    base_power = 50.0 * (0.5 + np.random.random())
                    
                    # Hardware factors
                    if hw == "cuda":
                        hw_factor = 5.0
                        power_factor = 3.0
                    elif hw == "webgpu":
                        hw_factor = 3.0
                        power_factor = 2.0
                    else:
                        hw_factor = 1.0
                        power_factor = 1.0
                    
                    # Batch size factors
                    batch_factor = np.sqrt(batch_size)
                    
                    # Calculate metrics
                    throughput = base_throughput * hw_factor * batch_factor * (1.0 + np.random.normal(0, 0.1))
                    latency = base_latency / hw_factor * (1.0 + 0.1 * batch_size) * (1.0 + np.random.normal(0, 0.1))
                    memory = base_memory * (1.0 + 0.2 * (batch_size - 1)) * (1.0 + np.random.normal(0, 0.05))
                    power = base_power * power_factor * (1.0 + 0.1 * batch_size) * (1.0 + np.random.normal(0, 0.1))
                    
                    # Add confidence and bounds
                    confidence = 0.85 + np.random.random() * 0.15
                    
                    # Calculate bounds
                    throughput_lower = throughput * (1.0 - (1.0 - confidence) * 2)
                    throughput_upper = throughput * (1.0 + (1.0 - confidence) * 2)
                    
                    # For time-series data
                    for timestamp in timestamps:
                        # Add time trend
                        time_position = timestamps.index(timestamp) / len(timestamps)
                        time_factor = 1.0 + 0.2 * np.sin(time_position * 2 * np.pi)
                        
                        data.append({
                            "model_name": model_name,
                            "model_category": model_category,
                            "hardware": hw,
                            "batch_size": batch_size,
                            "throughput": throughput * time_factor,
                            "latency_mean": latency * time_factor,
                            "memory_usage": memory * time_factor,
                            "power_consumption": power * time_factor,
                            "timestamp": timestamp.isoformat(),
                            "confidence": confidence,
                            "throughput_lower_bound": throughput_lower * time_factor,
                            "throughput_upper_bound": throughput_upper * time_factor,
                            "latency_lower_bound": latency / time_factor * 0.9,
                            "latency_upper_bound": latency / time_factor * 1.1,
                            "memory_lower_bound": memory * time_factor * 0.9,
                            "memory_upper_bound": memory * time_factor * 1.1
                        })
        
        return pd.DataFrame(data)
    
    def test_initialization(self):
        """Test initialization of visualization object."""
        self.assertEqual(self.vis.output_dir, str(self.output_dir))
        self.assertEqual(self.vis.interactive, False)
        self.assertEqual(self.vis.output_format, "png")
    
    def test_prepare_data(self):
        """Test data preparation functionality."""
        # Test with DataFrame
        df = self.vis._prepare_data(self.sample_data)
        self.assertIsInstance(df, pd.DataFrame)
        self.assertEqual(len(df), len(self.sample_data))
        
        # Test with dict
        data_dict = self.sample_data.to_dict('records')
        df_from_dict = self.vis._prepare_data(data_dict)
        self.assertIsInstance(df_from_dict, pd.DataFrame)
        
        # Test with JSON file
        json_path = self.output_dir / "test_data.json"
        with open(json_path, 'w') as f:
            json.dump(data_dict, f)
        df_from_json = self.vis._prepare_data(str(json_path))
        self.assertIsInstance(df_from_json, pd.DataFrame)
        
        # Test with CSV file
        csv_path = self.output_dir / "test_data.csv"
        self.sample_data.to_csv(csv_path, index=False)
        df_from_csv = self.vis._prepare_data(str(csv_path))
        self.assertIsInstance(df_from_csv, pd.DataFrame)
    
    def test_3d_visualization(self):
        """Test 3D visualization creation."""
        output_file = self.vis.create_3d_visualization(
            self.sample_data,
            x_metric="batch_size",
            y_metric="throughput",
            z_metric="memory_usage",
            color_metric="hardware",
            title="Test 3D Visualization"
        )
        
        # Check that output file exists
        self.assertTrue(os.path.exists(output_file))
        self.assertTrue(output_file.endswith(".png"))
    
    def test_performance_dashboard(self):
        """Test performance dashboard creation."""
        output_file = self.vis.create_performance_dashboard(
            self.sample_data,
            metrics=["throughput", "latency_mean"],
            groupby=["model_category", "hardware"],
            title="Test Performance Dashboard"
        )
        
        # Check that output file exists
        self.assertTrue(os.path.exists(output_file))
        self.assertTrue(output_file.endswith(".png"))
    
    def test_time_series_visualization(self):
        """Test time series visualization creation."""
        output_file = self.vis.create_time_series_visualization(
            self.sample_data,
            time_column="timestamp",
            metric="throughput",
            groupby=["model_name", "hardware"],
            title="Test Time Series"
        )
        
        # Check that output file exists
        self.assertTrue(os.path.exists(output_file))
        self.assertTrue(output_file.endswith(".png"))
    
    def test_power_efficiency_visualization(self):
        """Test power efficiency visualization creation."""
        output_file = self.vis.create_power_efficiency_visualization(
            self.sample_data,
            performance_metric="throughput",
            power_metric="power_consumption",
            groupby=["model_category"],
            title="Test Power Efficiency"
        )
        
        # Check that output file exists
        self.assertTrue(os.path.exists(output_file))
        self.assertTrue(output_file.endswith(".png"))
    
    def test_dimension_reduction_visualization(self):
        """Test dimension reduction visualization creation."""
        output_file = self.vis.create_dimension_reduction_visualization(
            self.sample_data,
            features=["batch_size", "memory_usage", "power_consumption", "latency_mean"],
            target="throughput",
            method="pca",
            n_components=2,
            groupby="model_category",
            title="Test Dimension Reduction"
        )
        
        # Check that output file exists
        self.assertTrue(os.path.exists(output_file))
        self.assertTrue(output_file.endswith(".png"))
    
    def test_prediction_confidence_visualization(self):
        """Test prediction confidence visualization creation."""
        output_file = self.vis.create_prediction_confidence_visualization(
            self.sample_data,
            metric="throughput",
            confidence_column="confidence",
            groupby=["model_category", "hardware"],
            title="Test Prediction Confidence"
        )
        
        # Check that output file exists
        self.assertTrue(os.path.exists(output_file))
        self.assertTrue(output_file.endswith(".png"))
    
    def test_batch_visualizations(self):
        """Test batch visualization creation."""
        visualization_files = self.vis.create_batch_visualizations(
            self.sample_data,
            metrics=["throughput", "latency_mean"],
            groupby=["model_category", "hardware"],
            include_3d=True,
            include_time_series=True,
            include_power_efficiency=True,
            include_dimension_reduction=True,
            include_confidence=True
        )
        
        # Check that visualization files dict is not empty
        self.assertIsInstance(visualization_files, dict)
        self.assertTrue(len(visualization_files) > 0)
        
        # Check that files exist
        for file_type, files in visualization_files.items():
            for file_path in files:
                self.assertTrue(os.path.exists(file_path))
    
    def test_visualization_report(self):
        """Test visualization report creation."""
        # Generate visualizations
        visualization_files = self.vis.create_batch_visualizations(
            self.sample_data,
            metrics=["throughput"],
            groupby=["model_category"],
            include_3d=True,
            include_time_series=True,
            include_power_efficiency=True,
            include_dimension_reduction=True,
            include_confidence=True
        )
        
        # Create report
        report_path = create_visualization_report(
            visualization_files=visualization_files,
            title="Test Visualization Report",
            output_file="test_report.html",
            output_dir=str(self.output_dir)
        )
        
        # Check that report exists
        self.assertTrue(os.path.exists(report_path))
        self.assertTrue(report_path.endswith(".html"))
        
        # Check report content
        with open(report_path, 'r') as f:
            content = f.read()
            self.assertIn("Test Visualization Report", content)
            self.assertIn("visualization-grid", content)

if __name__ == "__main__":
    unittest.main()